File: 01187. Make Array Strictly Increasing/solution.go
/**
 * Problem: 1187. Make Array Strictly Increasing
 *
 * Difficulty: Hard
 *
 * Language: Golang
 *
 * Performance: Runtime - 2 ms (Beats 100%)
 */

func makeArrayIncreasing(arr1 []int, arr2 []int) int {
	// Sort arr2 to enable binary search and deduplication
	sort.Ints(arr2)
	m := 0 // m will track the number of unique elements in arr2

	// Deduplicate arr2 in-place
	for _, x := range arr2 {
		if m == 0 || x != arr2[m-1] {
			arr2[m] = x
			m++
		}
	}

	// Resize arr2 to contain only unique elements
	arr2 = arr2[:m]
	const inf = 1 << 30 // Define a large constant to represent infinity

	// Add -inf at the beginning and inf at the end of arr1 for boundary handling
	arr1 = append([]int{-inf}, arr1...)
	arr1 = append(arr1, inf)

	n := len(arr1) // n is the length of the modified arr1
	f := make([]int, n) // f[i] stores the minimum operations to make arr1[0:i] strictly increasing

	// Initialize all values in f to infinity
	for i := range f {
		f[i] = inf
	}

	f[0] = 0 // No operation needed for the first element (-inf)

	// Dynamic programming to fill f
	for i := 1; i < n; i++ {
		// If previous element is less than current, carry forward the operation count
		if arr1[i-1] < arr1[i] {
			f[i] = f[i-1]
		}

		// Find the insertion point for arr1[i] in arr2 using binary search
		j := sort.SearchInts(arr2, arr1[i])

		// Try replacing up to k elements before i with elements from arr2
		for k := 1; k <= min(i-1, j); k++ {
			// Check if replacement maintains strictly increasing order
			if arr1[i-k-1] < arr2[j-k] {
				// Update f[i] with the minimum operations required
				f[i] = min(f[i], f[i-k-1]+k)
			}
		}
	}

	// If it's impossible to make arr1 strictly increasing, return -1
	if f[n-1] >= inf {
		return -1
	}

	// Return the minimum number of operations required
	return f[n-1]
}

File: 01187. Make Array Strictly Increasing/solution.js
/**
 * Problem: 1187. Make Array Strictly Increasing
 *
 * Difficulty: Hard
 *
 * Language: JavaScript
 *
 * Performance: Runtime - 19 ms (Beats 100%)
 */

/**
 * Returns the minimum operations to make arr1 strictly increasing using arr2
 *
 * @param {number[]} arr1 - Target array
 * @param {number[]} arr2 - Replacement options
 *
 * @returns {number} - Minimum operations or -1 if impossible
 */
const makeArrayIncreasing = (arr1, arr2) => {
  // Add sentinels to arr1 for easier boundary handling
  arr1 = [-1, ...arr1, Infinity]
  // Remove duplicates from arr2 and sort it in ascending order
  arr2 = [...new Set(arr2)].sort((a, b) => a - b)

  // Get the length of the modified arr1
  const n = arr1.length
  // Initialize dp array to track minimum operations for each position, filled with Infinity
  const dp = new Array(n).fill(Infinity)

  // Base case: zero operations needed for the first sentinel
  dp[0] = 0

  // Iterate through arr1 from the first real element to the end
  for (let i = 1; i < n; i++) {
    // Find the first index in arr2 where arr2[index] >= arr1[i]
    const j = search(arr2, arr1[i])

    // Try replacing up to k previous elements with arr2 values
    for (let k = 1; k <= Math.min(i - 1, j); k++)
      // Check if the replacement keeps the sequence strictly increasing
      if (arr1[i - k - 1] < arr2[j - k])
        // Update dp[i] with the minimum operations found
        dp[i] = Math.min(dp[i], dp[i - k - 1] + k)

    // Check if the current element is strictly greater than the previous
    if (arr1[i - 1] < arr1[i])
      // Update dp[i] if no replacement is needed
      dp[i] = Math.min(dp[i], dp[i - 1])
  }

  // Return the minimum operations if possible, otherwise -1
  return dp[n - 1] < Infinity ? dp[n - 1] : -1
}

/**
 * Binary search: returns first index where nums[index] >= target
 *
 * @param {number[]} nums - Sorted array
 * @param {number} target - Value to search for
 *
 * @returns {number} - Index in nums
 */
const search = (nums, target) => {
  // Initialize low and high pointers for binary search
  let [low, high] = [0, nums.length]

  // Perform binary search to find the first index where nums[index] >= target
  while (low < high) {
    // Calculate the middle index
    const mid = (low + high) >>> 1

    // If nums[mid] is greater than or equal to target, narrow search to left half
    // Otherwise, search in the right half
    nums[mid] >= target ? (high = mid) : (low = mid + 1)
  }

  // Return the index where nums[index] >= target or nums.length if not found
  return low
}
